### ISSUE TITLE
NullPointerException - CustomTabsSession is required for launching a TWA

### ISSUE THREAD

**Issue by tobiasharle** [GoogleChrome/android-browser-helper#38]
We recently launched our TWA application and saw a few of these crashes come in. We are using **0.1.0-alpha1**. I have not experienced this issue myself so I have no way of reproducing it.

```
java.lang.NullPointerException: CustomTabsSession is required for launching a TWA
    at androidx.browser.trusted.TrustedWebActivityIntentBuilder.build(TrustedWebActivityIntentBuilder.java:15)
    at com.google.androidbrowserhelper.trusted.TwaLauncher.launchWhenSplashScreenReady(TwaLauncher.java:2)
    at com.google.androidbrowserhelper.trusted.TwaLauncher.lambda$launchWhenSessionEstablished$2(TwaLauncher.java:1)
    at com.google.androidbrowserhelper.trusted.-$$Lambda$TwaLauncher$wy2yMaTyWzqDYy88T6GKCwq94sM.run
    at com.google.androidbrowserhelper.trusted.splashscreens.PwaWrapperSplashScreenStrategy.onSplashImageTransferred(PwaWrapperSplashScreenStrategy.java:2)
    at com.google.androidbrowserhelper.trusted.splashscreens.-$$Lambda$PwaWrapperSplashScreenStrategy$w9o5iqMEy62ZZ2i_sMAmRiQ7IGw.onFinished
    at com.google.androidbrowserhelper.trusted.splashscreens.SplashImageTransferTask$1.onPostExecute(SplashImageTransferTask.java:3)
    at android.os.AsyncTask.finish(AsyncTask.java:695)
    at android.os.AsyncTask.access$600(AsyncTask.java:180)
    at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:712)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:214)
    at android.app.ActivityThread.main(ActivityThread.java:7050)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:965)
```

**Comment by andreban** [comment]
@tobiasharle Would you be able to check the devices make and models?

@PEConn, FYI
 
It seems the issue is being caused by invoking this [line]( https://github.com/GoogleChrome/android-browser-helper/blob/0.1.0-alpha1/androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/TwaLauncher.java#L196) with a `null` session.

Interestingly, a check for `null` happens on `launchWhenSessionEstablished`, before invoking `launchWhenSplashScreenReady`, and that's the only 2 places the method are invoked.

**Comment by andreban** [comment]
So, I think the issue is happening when invoking from [this branch](https://github.com/GoogleChrome/android-browser-helper/blob/0.1.0-alpha1/androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/TwaLauncher.java#L186-L187).

It invokes [PwaWrapperSplashScreenStrategy](https://github.com/GoogleChrome/android-browser-helper/blob/0.1.0-alpha1/androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/splashscreens/PwaWrapperSplashScreenStrategy.java), which uses  [SplashImageTransferTask](https://github.com/GoogleChrome/android-browser-helper/blob/0.1.0-alpha1/androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/splashscreens/SplashImageTransferTask.java), which executes asynchronously. 

What may be happening is that the service is getting disconnected, [which sets `mSession` to null](https://github.com/GoogleChrome/android-browser-helper/blob/0.1.0-alpha1/androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/TwaLauncher.java#L254-L256).

**Comment by tobiasharle** [comment]
Sure, here are some stats on device models:
50.00% Samsung Galaxy S8
16.67% Samsung Galaxy S9
11.11% Samsung Galaxy S8+
11.11% Huawei P20 lite
5.56% Huawei P10
5.56% Samsung Galaxy J4+

**Comment by LJYq** [comment]
I also encountered the same problem, how can I solve it?

**Comment by PEConn** [comment]
Hey, I'm starting to look into this.

LJYq, are you able to reproduce this locally or is it something you're seeing in the wild?

**Comment by PEConn** [comment]
@tobiasharle, how often are you seeing these crashes? (It looks like you've got ~18 crashes, but how many users do you have, and do you know how many times they have successfully launched the TWA?)

We've tried to reproduce this and we can't - so we're not sure what a proper fix would be. It seems that the browser is disconnecting while we're trying to launch the TWA. I'm wondering what the best solution would be - I can see a few options, none of them are perfect:

1. Open the website as a Custom Tab. The user gets something usable but it's inconsistent from what they normally get.
2. Close the app. The user will probably re-open the app and chances are it will be successful this time.
3. Leave the loading screen there indefinitely. The user will probably quit the app and re-open it. (This is the worst option in my view.)
4. Let the app crash as it does now. From a user's point of view this is the same as option 2 (on later versions of Android at least), but as the developer you get visibility into how often it happens, whereas options 1, 2 and 3 would be invisible to you (until a user writes a poor app review).

I'm not sure it's possible to get a perfect "fix" for this - even if we could fix it in Chrome, this may happen for other apps (and our working theory is that this crash is caused by "Memory Saving" apps killing Chrome in the background).

**Comment by LJYq** [comment]
@PEConn 
Report here an exception about mSession is null,but we have already dealt with null, I do not know what happened?
TwaLauncher:
Intent intent = builder.build(mSession);
ContextCompat.startActivity(mContext, intent, null);

**Comment by tobiasharle** [comment]
Sorry for the delay..

The data above was reported early on after our release. More recently, the crash keeps happening about 100 times a day which is not a lot with around 20000 sessions daily.

I think I might try a 5:th option to catch when session is null and attempt to re-launch the twa again.
Hopefully that will then work and the user won't notice. Also add some logging to see if I can understand it better.

**Comment by pshmakov** [comment]
I'm quite sure this is just because the users close the app while seeing the splash screen. I landed the fix, let's see if it helps.

**Comment by tobiasharle** [comment]
Hi again, I have released an update with your fix included and sorry to say it hasn't helped. The issue is still at large.

I added a few logs as well to see what is happening and it shows me that destroy is NOT called when this issue occurs. For whatever reason, onServiceDisconnected is called (sometimes shortly) after onCustomTabsServiceConnected.

Let me know if there are any other logs I should add that could help solve this.

**Comment by rockeynebhwani** [comment]
We upgraded our TWA with android-browser-helper 1.0 stable version yesterday and we are also seeing similar errors. I will be able to provide number of errors and model etc when we have few days of data gathered in prod. 

java.lang.NullPointerException: 
  at androidx.browser.trusted.TrustedWebActivityIntentBuilder.build (TrustedWebActivityIntentBuilder.java:220)
  at com.google.androidbrowserhelper.trusted.TwaLauncher.launchWhenSplashScreenReady (TwaLauncher.java:229)
  at com.google.androidbrowserhelper.trusted.TwaLauncher.lambda$launchWhenSessionEstablished$3 (TwaLauncher.java:217)
  at com.google.androidbrowserhelper.trusted.-$$Lambda$TwaLauncher$Pgq1RErTrlHE2hqdSd7zUQR6Yz4.run (Unknown Source:6)
  at com.google.androidbrowserhelper.trusted.splashscreens.PwaWrapperSplashScreenStrategy.onSplashImageTransferred (PwaWrapperSplashScreenStrategy.java:202)
  at com.google.androidbrowserhelper.trusted.splashscreens.PwaWrapperSplashScreenStrategy.lambda$configureTwaBuilder$0 (PwaWrapperSplashScreenStrategy.java:195)
  at com.google.androidbrowserhelper.trusted.splashscreens.-$$Lambda$PwaWrapperSplashScreenStrategy$w9o5iqMEy62ZZ2i_sMAmRiQ7IGw.onFinished (Unknown Source:6)
  at com.google.androidbrowserhelper.trusted.splashscreens.SplashImageTransferTask$1.onPostExecute (SplashImageTransferTask.java:144)
  at com.google.androidbrowserhelper.trusted.splashscreens.SplashImageTransferTask$1.onPostExecute (SplashImageTransferTask.java:91)
  at android.os.AsyncTask.finish (AsyncTask.java:696)
  at android.os.AsyncTask.access$600 (AsyncTask.java:180)
  at android.os.AsyncTask$InternalHandler.handleMessage (AsyncTask.java:713)
  at android.os.Handler.dispatchMessage (Handler.java:112)
  at android.os.Looper.loop (Looper.java:216)
  at android.app.ActivityThread.main (ActivityThread.java:7625)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:524)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:987)

**Comment by andreban** [comment]
Just a quick update. We're trying to reproduce and looking into this. It seems the original issue was fixed, but the error is now happening in another place in the code.

**Comment by andreban** [comment]
@ankitarora05 I tested the APK on a Pixel 3 XL with Chrome 80, and the App is working. This makes me thing it's an issue related to a device / browser version. Can you confirm the device and browser combination you are seeing the crash on?

**Comment by andreban** [comment]
@rockeynebhwani We're looking into what's breaking.

**Comment by rockeynebhwani** [comment]
Thanks @andreban. Not sure if you needed details of device and browsers from me also.. attaching anyways

![image](https://user-images.githubusercontent.com/1718757/73939650-d82eab00-48e1-11ea-98f7-6a5433c26d8b.png)

**Comment by andreban** [comment]
@rockeynebhwani Thanks for sharing this. It does help indeed. It seems it only happens on devices either from Samsung or Huawei. @PEConn FYI

**Comment by PEConn** [comment]
That's interesting - maybe it's something to do with Samsung browser? What browser is the default for Huawei phones?

**Comment by andreban** [comment]
Linking to the issue on bugs.chromium.org: https://bugs.chromium.org/p/chromium/issues/detail?id=1052367

@rockeynebhwani would you have an updated list of devices where the crash is happening?

**Comment by rockeynebhwani** [comment]
@andreban  - Here is the latest device list (for last 60 days)

![image](https://user-images.githubusercontent.com/1718757/75595051-d2605b80-5a82-11ea-94d3-37533e8ecdcb.png)

**Comment by andreban** [comment]
@rockeynebhwani Thank you! I got a low-end galaxy device and trying to reproduce but with success so far.

**Comment by khmyznikov** [comment]
java.lang.NullPointerException
com.google.androidbrowserhelper.trusted.TwaLauncher.launchWhenSplashScreenReady

`java.lang.NullPointerException: 
  at androidx.browser.trusted.TrustedWebActivityIntentBuilder.build (TrustedWebActivityIntentBuilder.java:22)
  at com.google.androidbrowserhelper.trusted.TwaLauncher.launchWhenSplashScreenReady (TwaLauncher.java:22)
  at com.google.androidbrowserhelper.trusted.TwaLauncher.lambda$launchTwa$2 (TwaLauncher.java)
  at com.google.androidbrowserhelper.trusted.TwaLauncher.lambda$launchWhenSessionEstablished$3 (TwaLauncher.java)
  at com.google.androidbrowserhelper.trusted.-$$Lambda$TwaLauncher$Pgq1RErTrlHE2hqdSd7zUQR6Yz4.run (-.java:6)
  at com.google.androidbrowserhelper.trusted.splashscreens.PwaWrapperSplashScreenStrategy.onSplashImageTransferred (PwaWrapperSplashScreenStrategy.java:1)
  at com.google.androidbrowserhelper.trusted.splashscreens.PwaWrapperSplashScreenStrategy.lambda$configureTwaBuilder$0 (PwaWrapperSplashScreenStrategy.java:1)
  at com.google.androidbrowserhelper.trusted.splashscreens.-$$Lambda$PwaWrapperSplashScreenStrategy$w9o5iqMEy62ZZ2i_sMAmRiQ7IGw.onFinished (-.java:6)
  at com.google.androidbrowserhelper.trusted.splashscreens.SplashImageTransferTask$1.onPostExecute (SplashImageTransferTask.java:5)
  at com.google.androidbrowserhelper.trusted.splashscreens.SplashImageTransferTask$1.onPostExecute (SplashImageTransferTask.java:5)
  at android.os.AsyncTask.finish (AsyncTask.java:755)
  at android.os.AsyncTask.access$900 (AsyncTask.java:192)
  at android.os.AsyncTask$InternalHandler.handleMessage (AsyncTask.java:772)
  at android.os.Handler.dispatchMessage (Handler.java:107)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:7397)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:492)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:935)`

Nokia 6.1 (PL2_sprout), 2816MB RAM, Android 10

**Comment by andreban** [comment]
@khmyznikov are you able to reproduce this locally on this device? I've tried on a few devices, but this has been hard to repro so we can find the root cause and fix.

**Comment by khmyznikov** [comment]
@andreban It's from play console analytics :(

**Comment by andreban** [comment]
Thanks for confirming. I'll continue trying to repro and find the root cause.

**Comment by khmyznikov** [comment]
Same crash as i mention but on this device: HUAWEI P smart 2019 (HWPOT-H), 2816MB RAM, Android 9

**Comment by andreban** [comment]
#92 should address this. I'll update this issue when the next version is released

**Comment by AlexBorsody** [comment]
I was seeing this issue in the crash reports in March up into April but seems to be resolved now

**Comment by andreban** [comment]
This should have been fixed via #92. Please, re-open if using the latest version and still seeing the issue.
